src/macros-by-example.md

```diff
@@ -24,149 +24,404 @@
-rule in turn. It transcribes the first successful match. Matching and
-transcription are closely related to each other, and we will describe them
-together.
-
-The macro expander matches and transcribes every token that does not begin with
-a `$` literally, including delimiters. For parsing reasons, delimiters must be
-balanced, but they are otherwise not special.
-
-In the matcher, `$` _name_ `:` _designator_ matches the nonterminal in the Rust
-syntax named by _designator_. Valid designators are:
-
-* `item`: an [_Item_]
-* `block`: a [_BlockExpression_]
-* `stmt`: a [_Statement_] without the trailing semicolon
-* `pat`: a [_Pattern_]
-* `expr`: an [_Expression_]
-* `ty`: a [_Type_]
-* `ident`: an [IDENTIFIER_OR_KEYWORD]
-* `path`: a [_TypePath_] style path
-* `tt`: a [_TokenTree_]&nbsp;(a single [token] or tokens in matching delimiters `()`, `[]`, or `{}`)
-* `meta`: a [_MetaItem_], the contents of an attribute
-* `lifetime`: a [LIFETIME_TOKEN]
-* `vis`: a [_Visibility_] qualifier
-* `literal`: matches `-`<sup>?</sup>[_LiteralExpression_]
+Each macro by example has a name, and one or more _rules_. Each ruleh as two
+parts: a _matcher_, describing the syntax that it matches, and a _transcriber_,
+describing the syntax that will replace a successfully matched invocation. Both
+the matcher and the transcriber must be surrounded by delimiters. Macros can
+expand to expressions, statements, items, types, or patterns.
+
+When a macro is invoked, the macro expander looks up macro invocations by name,
+and tries each macro rule in turn. It transcribes the first successful match; if
+this results in an error, then future matches are not tried. When matching, no
+lookahead is performed; if the compiler cannot umambiguously determine how to
+parse the macro invocation one token at a time, then it is an error. In the
+following example, the compiler does not look ahead past the identifier to see
+if the following token is a `)`, even though that would allow it to parse the
+invocation unambiguously:
+
+```compile_fail
+macro_rules! ambiguity {
+    ($($i:ident)* $j:ident) => { ($($i)-*) * $j };
+}
+
+ambiguity!(error); // Error: local ambiguity
+```
+
+In both the matcher and the transcriber, the `$` token is used to invoke special
+behaviours from the macro engine. Tokens that aren't part of such an invocation
+are matched and transcribed literally, with one exception. The exception is that
+the outer delimiters for the matcher will match any pair of delimiters. Thus,
+for instance, the matcher `(())` will match `{()}` but not `{{}}`. The character
+`$` cannot be matched or transcribed literally.
+
+## Metavariables
+
+In the matcher, `$` _name_ `:` _fragment-specifier_ matches a Rust syntax
+fragment of the kind specified and binds it to the metavariable `$`_name_. Valid
+fragment specifiers are:
+
+  * `item`: an [_Item_]
+  * `block`: a [_BlockExpression_]
+  * `stmt`: a [_Statement_] without the trailing semicolon
```

Comment:
Unless it's an item statement and the trailing semicolon is a part of the item syntax.